crates/sui-adapter/src/programmable_transactions/linkage_view.rs

```diff
@@ -0,0 +1,258 @@
+            .get(struct_)
+            .cloned()
+    }
+
+    fn add_type_origin(
+        &self,
+        runtime_id: ModuleId,
+        struct_: Identifier,
+        defining_id: ObjectID,
+    ) -> Result<(), ExecutionError> {
+        let mut cache = self.type_origin_cache.borrow_mut();
+        let module_cache = cache.entry(runtime_id.clone()).or_default();
+
+        match module_cache.entry(struct_) {
+            Entry::Vacant(entry) => {
+                entry.insert(*defining_id);
+            }
+
+            Entry::Occupied(entry) => {
+                if entry.get() != &*defining_id {
+                    return Err(ExecutionError::invariant_violation(format!(
+                        "Conflicting defining ID for {}::{}: {} and {}",
+                        runtime_id,
+                        entry.key(),
+                        defining_id,
+                        entry.get(),
+                    )));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl From<&MovePackage> for LinkageInfo {
+    fn from(package: &MovePackage) -> Self {
+        Self {
+            link_context: package.id().into(),
+            original_package_id: package.original_package_id().into(),
+            linkage_table: package.linkage_table().clone(),
+        }
+    }
+}
+
+impl<'state, S: StorageView> LinkageResolver for LinkageView<'state, S> {
+    type Error = SuiError;
+
+    fn link_context(&self) -> AccountAddress {
+        if let Some(LinkageInfo { link_context, .. }) = &self.linkage_info {
+            *link_context
+        } else {
+            AccountAddress::ZERO
+        }
+    }
+
+    fn relocate(&self, module_id: &ModuleId) -> Result<ModuleId, Self::Error> {
+        let Some(linkage) = &self.linkage_info else {
+            return Ok(module_id.clone());
+        };
+
+        // The request is to relocate a module in the package that the link context is from.  This
+        // entry will not be stored in the linkage table, so must be handled specially.
+        if module_id.address() == &linkage.original_package_id {
+            return Ok(ModuleId::new(
+                linkage.link_context,
+                module_id.name().to_owned(),
+            ));
+        }
+
+        let runtime_id = ObjectID::from_address(*module_id.address());
+        let Some(upgrade) = linkage.linkage_table.get(&runtime_id) else {
+            return Err(ExecutionError::invariant_violation(format!(
+                "Missing linkage for {runtime_id} in context {}",
+                linkage.link_context,
+            )).into());
+        };
+
+        Ok(ModuleId::new(
+            upgrade.upgraded_id.into(),
+            module_id.name().to_owned(),
+        ))
+    }
+
+    fn defining_module(
+        &self,
+        runtime_id: &ModuleId,
+        struct_: &IdentStr,
+    ) -> Result<ModuleId, Self::Error> {
+        if self.linkage_info.is_none() {
+            return Ok(runtime_id.clone());
```

Comment:
Can this potentially lead to a silent error, for example, if we forget to set the context appropriately?

Reply:
Yes, that can happen.  Currently there are two states for the linkage info -- `None` as in "off" which does no re-linking, or `Some(info)` which means re-link using that information.  We use `None` to sigal both that we haven't set a linkage when we might have wanted to, and that we intentionally set no linkage because upgrades have been disabled.

I'll do some hardening which will split this out into three different states, where linkage requests when no linkage is provided will error, and it will also be an error to overwrite an existing linkage that hasn't been reset.